test/relayHub/RelayHub.test.ts

```diff
@@ -0,0 +1,1199 @@
+        let chainId: number;
+        let relayHub: string;
+        let penalizer: PenalizerInstance;
+        let relayHubInstance: RelayHubInstance;
+        let recipientContract: TestRecipientInstance;
+        let verifierContract: TestVerifierEverythingAcceptedInstance;
+        let deployVerifierContract: TestDeployVerifierEverythingAcceptedInstance;
+        let forwarderInstance: IForwarderInstance;
+        let target: string;
+        let verifier: string;
+        let forwarder: string;
+        let gaslessAccount: AccountKeypair;
+        const gasLimit = '3000000';
+        const gasPrice = '1';
+        let sharedRelayRequestData: RelayRequest;
+        let sharedDeployRequestData: DeployRequest;
+
+        let env: Environment;
+        let token: TestTokenInstance;
+        let factory: SmartWalletFactoryInstance;
+        let relayRequest: RelayRequest;
+
+        let misbehavingVerifier: TestDeployVerifierConfigurableMisbehaviorInstance;
+        let signatureWithMisbehavingVerifier: string;
+        let relayRequestMisbehavingVerifier: DeployRequest;
+
+        describe('Manager tests - Stake related scenarios', async () => {
+            beforeEach(async function () {
+                penalizer = await Penalizer.new();
+                relayHubInstance = await deployHub(penalizer.address);
+            });
+
+            it('Should stake only from owner account', async () => {
+                await relayHubInstance.stakeForAddress(relayManager, 1000, {
+                    value: ether('1'),
+                    from: relayOwner
+                });
+                await assert.isRejected(
+                    relayHubInstance.stakeForAddress(relayManager, 1000, {
+                        value: ether('1'),
+                        from: _
+                    }),
+                    'not owner',
+                    'Stake was not made by the owner account'
+                );
+            });
+
+            it('Should NOT be able to unauthorize/unstake a HUB and then perform a new stake with a worker', async () => {
+                await relayHubInstance.stakeForAddress(relayManager, 1000, {
+                    value: ether('1'),
+                    from: relayOwner
+                });
+
+                await relayHubInstance.addRelayWorkers([relayWorker], {
+                    from: relayManager
+                });
+                await relayHubInstance.unlockStake(relayManager, {
+                    from: relayOwner
+                });
+
+                //Verifying stake is now unlocked
+                const stakeInfo = await relayHubInstance.getStakeInfo(
+                    relayManager
+                );
+                const isUnlocked = Number(stakeInfo.withdrawBlock) > 0;
+                assert.isTrue(isUnlocked, 'Stake is not unlocked');
+
+                //Moving blocks to be able to unstake
+                await evmMineMany(Number(stakeInfo.unstakeDelay));
```

Comment:
We could try to do this async or show a message that would tike a little bit longer.

Reply:
Do you mean we shouldn't wait for the blocks to be mined? If so I don't think we could avoid this. Before allowing the owner to retrieve the stake, they need to wait for a certain number of blocks to be mined, so here we are simulating exactly this.
Please have a look at the [stakeForAddress function](https://github.com/rsksmart/rif-relay-contracts/blob/master/contracts/RelayHub.sol#L312).

Reply:
What I was referring to, its instead of using an async function we could just execute the evmMineMany and keep with the flow. I have seen some tests that do that.

Reply:
Talked with Antonio about this change